tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -1089,6 +1091,91 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
   return success();
 }
 
+LogicalResult ConvertTFLConv3DOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_conv3d_op = cast<TFL::Conv3DOp>(op);
+  RankedTensorType input_type =
+      tfl_conv3d_op.input().getType().dyn_cast<RankedTensorType>();
+  RankedTensorType filter_type =
+      tfl_conv3d_op.filter().getType().dyn_cast<RankedTensorType>();
+  ShapedType output_type =
+      tfl_conv3d_op.getResult().getType().dyn_cast<ShapedType>();
+
+  // Not a ranked tensor output
+  if (!input_type) return failure();
+  if (!output_type) return failure();
+  if (!filter_type) return failure();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool filter_is_qtype =
+      filter_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::QuantizedType>();
+
+  if ((input_is_qtype != filter_is_qtype) ||
+      (input_is_qtype != output_is_qtype)) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLConv3DOp: input/filter/output tensor should "
```

Comment:
Update to remove ConvertTFLConv3DOp and trailing punctuation?

Reply:
I didn't notice MLIR error convention at that time. Removed, will never again.